### PR TITLE
[FEATURE] Introduce Pre and Post rendering document events

### PR DESCRIPTION
**Pull request by linawolf**
I am using the pre event to pass additional information, for example the menu structure to my template. The post event could be used for post processing for example.

### CODE REVIEW

File: docs/en/events.rst [hunk 1]
```diff
@@ -79,3 +79,11 @@ The events you can listen for are as follows:
   in implementing projects or extending packages. See
   `Package Intersphinx, provided by the TYPO3 Documentation Team <https://github.com/TYPO3-Documentation/intersphinx>`__
   for an example implementation.
+- ``PreDocumentRenderEvent``: This event is called in the DocumentNodeRenderer
```
Comment (greg0ire) [inline docs/en/events.rst]: I'd either use "in `DocumentNodeRendererer`" (with code markup), or "in the document node renderer" here and in similar cases below.

File: lib/Event/PostDocumentRenderEvent.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+use Doctrine\RST\Configuration;
+use Doctrine\RST\Nodes\DocumentNode;
+use Doctrine\RST\Templates\TemplateRenderer;
+
+use function sprintf;
+
+/**
+ * This event is called in the DocumentNodeRenderer after the content of
+ * the document got rendered and before the HTML is returned. It can be used
+ * to postprocess or exchange the HTML that will be written to the file.
```
Comment (greg0ire) [inline lib/Event/PostDocumentRenderEvent.php]: Did you really mean to write "exchange"? To me, you exchange something with something else.

```suggestion
 * to postprocess or change the HTML that will be written to the file.
```

**Review by greg0ire — APPROVED** [review]

File: docs/en/events.rst (hunk 1 above)
Comment (linawolf) [inline docs/en/events.rst]:
```suggestion
- ``PreDocumentRenderEvent``: This event is called in the `DocumentNodeRenderer`
```

File: docs/en/events.rst [hunk 3]
```diff
@@ -79,3 +79,11 @@ The events you can listen for are as follows:
   in implementing projects or extending packages. See
   `Package Intersphinx, provided by the TYPO3 Documentation Team <https://github.com/TYPO3-Documentation/intersphinx>`__
   for an example implementation.
+- ``PreDocumentRenderEvent``: This event is called in the `DocumentNodeRenderer`
+  before the document Node is rendered. The event can be used to influence the
+  parameters sent to the template and or the name of the template to be used
+  during rendering.
+- ``PostDocumentRenderEvent``: This event is called in the DocumentNodeRenderer
```
Comment (linawolf) [inline docs/en/events.rst]:
```suggestion
- ``PostDocumentRenderEvent``: This event is called in the `DocumentNodeRenderer`
```

File: docs/en/events.rst [hunk 4]
```diff
@@ -79,3 +79,11 @@ The events you can listen for are as follows:
   in implementing projects or extending packages. See
   `Package Intersphinx, provided by the TYPO3 Documentation Team <https://github.com/TYPO3-Documentation/intersphinx>`__
   for an example implementation.
+- ``PreDocumentRenderEvent``: This event is called in the `DocumentNodeRenderer`
```
Comment (greg0ire) [inline docs/en/events.rst:82]: Shouldn't you be using double backticks? Since this isn't markdown?

Comment (linawolf) [reply docs/en/events.rst:82]: As our default textrole is "code" the result is exactly the same: https://github.com/doctrine/rst-parser/pull/234

Also GiThub renders it like code as well. The only advantage of double backticks then is that you can use single backticks within without escaping them

**Review by SenseException — APPROVED** [review]

**Pull request merged**